core/src/main/java/com/linecorp/armeria/common/stream/FuseableStreamMessage.java

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.linecorp.armeria.common.stream.StreamMessageUtil.EMPTY_OPTIONS;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.stream.FuseableStreamMessage.MapperFunction.Type;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class FuseableStreamMessage<T, U> implements StreamMessage<U> {
+
+    static <T> FuseableStreamMessage<T, T> of(StreamMessage<? extends T> source,
+                                              Predicate<? super T> predicate) {
+        return new FuseableStreamMessage<>(source, MapperFunction.of(predicate));
+    }
+
+    static <T, R> FuseableStreamMessage<T, R> of(StreamMessage<? extends T> source,
+                                                 Function<? super T, ? extends R> function) {
+        return new FuseableStreamMessage<>(source, MapperFunction.of(function));
+    }
+
+    private final StreamMessage<Object> source;
+    private final List<MapperFunction<Object, Object>> functions;
```

Comment:
Would it be a good idea to chain `MapperFunction`s instead of using a `List`? :thinking:

Reply:
It is possible. That looks the type signature clean.